shell/platform/android/io/flutter/embedding/engine/loader/FlutterLoader.java

```diff
@@ -0,0 +1,342 @@
+// found in the LICENSE file.
+
+package io.flutter.embedding.engine.loader;
+
+import android.content.Context;
+import android.content.pm.ApplicationInfo;
+import android.content.pm.PackageManager;
+import android.content.res.AssetManager;
+import android.os.Bundle;
+import android.os.Handler;
+import android.os.Looper;
+import android.os.SystemClock;
+import android.support.annotation.NonNull;
+import android.support.annotation.Nullable;
+import android.util.Log;
+import android.view.WindowManager;
+
+import io.flutter.BuildConfig;
+import io.flutter.embedding.engine.FlutterJNI;
+import io.flutter.util.PathUtils;
+import io.flutter.view.VsyncWaiter;
+
+import java.io.File;
+import java.util.*;
+
+/**
+ * This class finds Flutter resources in an application APK and also loads Flutter's native library.
+ */
+public class FlutterLoader {
+    private static final String TAG = "FlutterLoader";
+
+    // Must match values in flutter::switches
+    private static final String AOT_SHARED_LIBRARY_NAME = "aot-shared-library-name";
+    private static final String SNAPSHOT_ASSET_PATH_KEY = "snapshot-asset-path";
+    private static final String VM_SNAPSHOT_DATA_KEY = "vm-snapshot-data";
+    private static final String ISOLATE_SNAPSHOT_DATA_KEY = "isolate-snapshot-data";
+    private static final String FLUTTER_ASSETS_DIR_KEY = "flutter-assets-dir";
+
+    // XML Attribute keys supported in AndroidManifest.xml
+    public static final String PUBLIC_AOT_SHARED_LIBRARY_NAME =
+        FlutterLoader.class.getName() + '.' + AOT_SHARED_LIBRARY_NAME;
+    public static final String PUBLIC_VM_SNAPSHOT_DATA_KEY =
+        FlutterLoader.class.getName() + '.' + VM_SNAPSHOT_DATA_KEY;
+    public static final String PUBLIC_ISOLATE_SNAPSHOT_DATA_KEY =
+        FlutterLoader.class.getName() + '.' + ISOLATE_SNAPSHOT_DATA_KEY;
+    public static final String PUBLIC_FLUTTER_ASSETS_DIR_KEY =
+        FlutterLoader.class.getName() + '.' + FLUTTER_ASSETS_DIR_KEY;
+
+    // Resource names used for components of the precompiled snapshot.
+    private static final String DEFAULT_AOT_SHARED_LIBRARY_NAME = "libapp.so";
+    private static final String DEFAULT_VM_SNAPSHOT_DATA = "vm_snapshot_data";
+    private static final String DEFAULT_ISOLATE_SNAPSHOT_DATA = "isolate_snapshot_data";
+    private static final String DEFAULT_LIBRARY = "libflutter.so";
+    private static final String DEFAULT_KERNEL_BLOB = "kernel_blob.bin";
+    private static final String DEFAULT_FLUTTER_ASSETS_DIR = "flutter_assets";
+
+    // Mutable because default values can be overridden via config properties
+    private String aotSharedLibraryName = DEFAULT_AOT_SHARED_LIBRARY_NAME;
+    private String vmSnapshotData = DEFAULT_VM_SNAPSHOT_DATA;
+    private String isolateSnapshotData = DEFAULT_ISOLATE_SNAPSHOT_DATA;
+    private String flutterAssetsDir = DEFAULT_FLUTTER_ASSETS_DIR;
+
+    private static FlutterLoader instance;
+
+    @NonNull
+    public static FlutterLoader getInstance() {
```

Comment:
nit: Docs here would be useful too. I would maybe rename it to "getOrCreate" just to prevent any confusion about how to set the isntance.

Reply:
Added docs. And I hear what you're saying about `getOrCreate()`. I'm going to opt to leave it as-is to match other `getInstance()`s in the codebase.